temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java

```diff
@@ -798,16 +819,106 @@ public void respondNexusTaskFailed(
       RespondNexusTaskFailedRequest request,
       StreamObserver<RespondNexusTaskFailedResponse> responseObserver) {
     try {
-      NexusTaskToken taskToken = NexusTaskToken.fromBytes(request.getTaskToken());
-      TestWorkflowMutableState mutableState = getMutableState(taskToken.getExecutionId());
-      mutableState.failNexusTask(taskToken.getScheduledEventId(), request);
+      if (!request.hasError()) {
+        throw Status.INVALID_ARGUMENT
+            .withDescription("Nexus handler error not set on RespondNexusTaskFailedRequest")
+            .asRuntimeException();
+      }
+      NexusOperationRef ref = NexusOperationRef.fromBytes(request.getTaskToken());
+      TestWorkflowMutableState mutableState = getMutableState(ref.getExecutionId());
+      Failure failure = handlerErrorToFailure(request.getError());
+      mutableState.failNexusOperation(ref, failure);
       responseObserver.onNext(RespondNexusTaskFailedResponse.getDefaultInstance());
       responseObserver.onCompleted();
     } catch (StatusRuntimeException e) {
       handleStatusRuntimeException(e, responseObserver);
     }
   }
 
+  public void completeNexusOperation(ByteString token, HistoryEvent completionEvent) {
+    NexusOperationRef ref = NexusOperationRef.fromBytes(token);
+    TestWorkflowMutableState target = getMutableState(ref.getExecutionId());
+
+    switch (completionEvent.getEventType()) {
+      case EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+        Payloads result =
+            completionEvent.getWorkflowExecutionCompletedEventAttributes().getResult();
+        // All of our SDKs support returning a single value from workflows, we can safely ignore the
+        // rest of the payloads. Additionally, even if a workflow could return more than a single
+        // value,
+        // Nexus does not support it.
+        Payload p =
+            (result.getPayloadsCount() > 0) ? result.getPayloads(0) : Payload.getDefaultInstance();
+        target.completeNexusOperation(ref, p);
+        break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_FAILED:
+        target.failNexusOperation(
+            ref, completionEvent.getWorkflowExecutionFailedEventAttributes().getFailure());
```

Comment:
In the real server, we translate this failure to a minimal representation, stripping out the causes.
This may change in the real server but I would stay consistent for now.